### ISSUE TITLE
TimeSeriesFeaturizer doesn't appear to use correct significant lags in most cases

### ISSUE THREAD

**Issue by thehomebrewnerd** [alteryx/evalml#3770]
The lagged features that get generated by the `TimeSeriesFeaturizer` do not use the significant lags found by `_find_significant_lags` in most cases. Changing the forecast horizon and/or the gap parameters change the lagged features that are created, but these parameters are not directly used when identifying which lags are significant, only the max_delay is used.

The current process is something like this, if I am following the code correctly:
- Find all significant lags in the range of (0, max_delay)
- Add the start delay (which is gap + forecast_horizon) to the lags found from the previous step

This means that the significant lags that were identified can be offset, making them insignificant.

I believe the proper approach should be more like this:
- Use the gap and forecast_horizon and max_delay to identify the rows that fall within the feature engineering window
- Identify significant lags that fall within the feature engineering window
- Create lagged features from those significant lags without further modification

The problem is illustrated in the example below. Changing the forecast horizon changes the lagged features that are created.

**Forecast Horizon of 1**

```python
import pandas as pd
import featuretools as ft

from evalml.pipelines import TimeSeriesRegressionPipeline

df = pd.read_csv("DailyDelhiClimateTrain.csv", nrows=200)

max_delay = 10
gap = 0
forecast_horizon = 1

component_graph_as_list = ["Time Series Featurizer"]
parameters = {
    "pipeline": {
        "time_index": "date",
        "gap": gap,
        "max_delay": max_delay,
        "forecast_horizon": forecast_horizon,
    },
    "Time Series Featurizer": {
        "time_index": "date",
        "gap": gap,
        "max_delay": max_delay,
        "forecast_horizon": forecast_horizon,
    }
}
pipeline = TimeSeriesRegressionPipeline(component_graph=component_graph_as_list, parameters=parameters)

y = df.pop("meantemp")
X = df[["date"]]

p = pipeline.fit(X, y)

# For simplicity assume the first and fourth lags were found to be significant (zero indexed)
p.component_graph["Time Series Featurizer"].statistically_significant_lags = [0, 3]

p.transform(X, y).head()
```

For this first case, we create target delay features of 1 period and 4 periods as we expect
```
        date  target_delay_1  target_delay_4  target_rolling_mean
0 2013-01-01             NaN             NaN                  NaN
1 2013-01-02       10.000000             NaN                  NaN
2 2013-01-03        7.400000             NaN                  NaN
3 2013-01-04        7.166667             NaN             8.188889
4 2013-01-05        8.666667            10.0             7.744444
```

If we change nothing but making `forecast_horizon = 2` and run this again, we get a different result. We now create delay features of 2 period and 5 periods, neither of which were identified as significant. It seems to me that changing the number of points we want to forecast in the future should have no bearing on which lag features are created, only on which lags we can use.

For this case, we could not use the lag of 1 feature, because we would have no data available for the second forecast point (the feature engineering window doesn't include the previous row in this case). We should still be able to use the lag of 4 feature, which was significant.

**Forecast Horizon of 2**
```
        date  target_delay_2  target_delay_5  target_rolling_mean
0 2013-01-01             NaN             NaN                  NaN
1 2013-01-02             NaN             NaN                  NaN
2 2013-01-03       10.000000             NaN                  NaN
3 2013-01-04        7.400000             NaN                  NaN
4 2013-01-05        7.166667             NaN             8.188889
```

**Comment by dsherry** [comment]
Thanks for filing @thehomebrewnerd 
Thanks for posting a reproducer and sharing the feature matrices. I think this is a clear way to communicate and understand problems like this.
Nice trick with manually setting `statistically_significant_lags`. I thought you’d have to use mocking in order to do that.

In your discussion of the current process and proper approach, I suggest you call out which row you’re treating as having a lag of 0. To me, this should be the latest row in the feature engineering window. I mention it here because while the current evalml code indexes relative to this point during the computation of the lagged features, we then name them instead relative to the point of prediction. In other words, for a gap of 7, the feature “Feature_1_delay_8” in the final feature matrix corresponds to 8 units back from the prediction point, which is 1 row back from the edge of the feature engineering window.

> This means that the significant lags that were identified can be offset, making them insignificant.
Is another way to say this that:
The code which computes the delay values of the most significant lags, and the code which applies those lag amounts to compute the feature values themselves, are using different indexes, when they should actually be the same?

In other words, we are generating features with the wrong lagged amounts. 


I also have one question, mainly to understand the impact of this bug: ignoring the question of which lag amounts are optimal, is the feature matrix itself correct here? When we label a feature as a lag of 2, do the values in the feature matrix actually match with a lag of 2?

**Comment by thehomebrewnerd** [comment]
@dsherry 

1. I'm not sure I understand your question here exactly. We might need to discuss this more.
2. Yes, I believe we are generating features with the wrong lagged amounts. We identify which lags are significant and then we modify those values when creating lagged features. I believe it would make the most sense for the `_find_significant_lags` function to return the actual significant lags  that should be directly used to create lagged features without further adjustment. It's hard to follow what is happening when we call a function to get significant values, then proceed to modify those values before using them.
3. Yes. If you concat the target column to the feature matrix, you will see the lags match as you expect:

```
        date  target_delay_2  target_delay_5  target_rolling_mean   meantemp
0 2013-01-01             NaN             NaN                  NaN  10.000000
1 2013-01-02             NaN             NaN                  NaN   7.400000
2 2013-01-03       10.000000             NaN                  NaN   7.166667
3 2013-01-04        7.400000             NaN                  NaN   8.666667
4 2013-01-05        7.166667             NaN             8.188889   6.000000
5 2013-01-06        8.666667       10.000000             7.744444   7.000000
6 2013-01-07        6.000000        7.400000             7.277778   7.000000
7 2013-01-08        7.000000        7.166667             7.222222   8.857143
```

**Comment by jeremyliweishih** [comment]
After speaking with @thehomebrewnerd - [heres](https://github.com/alteryx/evalml/pull/3773) a potential path forward.

FYI @dsherry
